File: dotnet/DesignPatterns/BehavioralDesignPatterns/MementoDesignPattern/MementoDesignPattern.Example1/TextEditor.cs
namespace MementoDesignPattern.Example1;

/// <summary>
/// Originator: The object whose internal state changes over time and needs to be saved/restored.
/// TextEditor holds the document content, font settings, and cursor position.
/// It can produce a Memento (DocumentSnapshot) capturing its current state via CreateSnapshot(),
/// and restore a previous state from a given Memento via RestoreFromSnapshot().
/// </summary>
public class TextEditor 
{
    public string DocumentId { get; }
    public string Content { get; private set; }
    public string FontName { get; private set; }
    public int FontSize { get; private set; }
    public int CursorPosition { get; private set; }

    public TextEditor(string documentId) 
    {
        DocumentId = documentId;
        Content = string.Empty;
        FontName = "Arial";
        FontSize = 12;
        CursorPosition = 0;
    }

    #region Methods to modify the document

    public void Type(string text)
    {
        Content += text;
        CursorPosition += text.Length;
        Console.WriteLine($"[{DocumentId}] Typed '{text}' --> Content: {Content} | Cursor Position: {CursorPosition}");

    }

    public void ChangeFont(string fontName, int fontSize)
    {
        FontName = fontName;
        FontSize = fontSize;
        Console.WriteLine($"[{DocumentId}] Font changed to {FontName} with size {FontSize}");
    }

    public void MoveCursor(int position)
    {
        CursorPosition = position;
        Console.WriteLine($"[{DocumentId}] Cursor moved to position {CursorPosition}");
    }
    #endregion

    #region Memento-related methods
    public DocumentSnapshot CreateSnapshot(string name = "Auto-Save")
    {
        return new DocumentSnapshot(Content, FontName, FontSize, CursorPosition, name);
    }

    public void RestoreFromSnapshot(DocumentSnapshot documentSnapshot)
    {
        if (documentSnapshot is null) return;

        Content = documentSnapshot.Content;
        FontName = documentSnapshot.FontName;
        FontSize = documentSnapshot.FontSize;
        CursorPosition = documentSnapshot.CursorPosition;

        Console.WriteLine($"[{DocumentId}] Document restored from snapshot '{documentSnapshot.SnapshotName}' ({documentSnapshot.CreatedAt})");
    }

    public void DisplayStatus()
    {
        Console.WriteLine($"[{DocumentId}] Content: '{Content}' | Font: {FontName} ({FontSize}pt) | Cursor Position: {CursorPosition}");
    } 
    #endregion
}

File: nodejs/DesignPatterns/BehavioralDesignPatterns/MementoDesignPattern/MementoDesignPattern.Example1/src/TextEditor.ts
import { DocumentSnapshot } from './DocumentSnapshot';

export class TextEditor {
  readonly documentId: string;
  private _content: string;
  private _fontName: string;
  private _fontSize: number;
  private _cursorPosition: number;

  get content(): string { return this._content; }
  get fontName(): string { return this._fontName; }
  get fontSize(): number { return this._fontSize; }
  get cursorPosition(): number { return this._cursorPosition; }

  constructor(documentId: string) {
    this.documentId = documentId;
    this._content = '';
    this._fontName = 'Arial';
    this._fontSize = 12;
    this._cursorPosition = 0;
  }

  type(text: string): void {
    this._content += text;
    this._cursorPosition += text.length;
    console.log(`[${this.documentId}] Typed '${text}' --> Content: ${this._content} | Cursor Position: ${this._cursorPosition}`);
  }

  changeFont(fontName: string, fontSize: number): void {
    this._fontName = fontName;
    this._fontSize = fontSize;
    console.log(`[${this.documentId}] Font changed to ${this._fontName} with size ${this._fontSize}`);
  }

  moveCursor(position: number): void {
    this._cursorPosition = position;
    console.log(`[${this.documentId}] Cursor moved to position ${this._cursorPosition}`);
  }

  createSnapshot(name: string = 'Auto-Save'): DocumentSnapshot {
    return new DocumentSnapshot(this._content, this._fontName, this._fontSize, this._cursorPosition, name);
  }

  restoreFromSnapshot(snapshot: DocumentSnapshot): void {
    this._content = snapshot.content;
    this._fontName = snapshot.fontName;
    this._fontSize = snapshot.fontSize;
    this._cursorPosition = snapshot.cursorPosition;
    console.log(`[${this.documentId}] Document restored from snapshot '${snapshot.snapshotName}' (${snapshot.createdAt.toISOString()})`);
  }

  displayStatus(): void {
    console.log(`[${this.documentId}] Content: '${this._content}' | Font: ${this._fontName} (${this._fontSize}pt) | Cursor Position: ${this._cursorPosition}`);
  }
}
